sql/src/main/java/io/crate/executor/transport/TransportFetchNodeAction.java

```diff
@@ -0,0 +1,221 @@
+import org.elasticsearch.transport.TransportChannel;
+import org.elasticsearch.transport.TransportService;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.List;
+import java.util.UUID;
+
+@Singleton
+public class TransportFetchNodeAction {
+
+    private final String transportAction = "crate/sql/node/fetch";
+    private final TransportService transportService;
+    private final ClusterService clusterService;
+    private final StatsTables statsTables;
+    private final CircuitBreaker circuitBreaker;
+    private final CollectContextService collectContextService;
+    private final String executorName = ThreadPool.Names.SEARCH;
+    private final ThreadPool threadPool;
+    private final Functions functions;
+
+    @Inject
+    public TransportFetchNodeAction(TransportService transportService,
+                                    ThreadPool threadPool,
+                                    ClusterService clusterService,
+                                    StatsTables statsTables,
+                                    Functions functions,
+                                    CircuitBreakerService breakerService,
+                                    CollectContextService collectContextService) {
+        this.transportService = transportService;
+        this.clusterService = clusterService;
+        this.statsTables = statsTables;
+        this.circuitBreaker = breakerService.getBreaker(CrateCircuitBreakerService.QUERY_BREAKER);
+        this.collectContextService = collectContextService;
+        this.threadPool = threadPool;
+        this.functions = functions;
+
+        transportService.registerHandler(transportAction, new TransportHandler());
+    }
+
+    public void execute(
+            String targetNode,
+            NodeFetchRequest request,
+            ActionListener<NodeFetchResponse> listener) {
+        new AsyncAction(targetNode, request, listener).start();
+    }
+
+    private void nodeOperation(final NodeFetchRequest request,
+                               final ActionListener<NodeFetchResponse> fetchResponse) {
+        final UUID operationId = UUID.randomUUID();
+        statsTables.operationStarted(operationId, request.jobId(), "fetch");
+        String ramAccountingContextId = String.format("%s: %s", request.jobId(), operationId);
+        final RamAccountingContext ramAccountingContext = new RamAccountingContext(ramAccountingContextId, circuitBreaker);
+
+        NodeFetchOperation fetchOperation = new NodeFetchOperation(
+                request.jobId(),
+                request.jobSearchContextDocIds(),
+                request.toFetchReferences(),
+                request.closeContext(),
+                collectContextService,
+                threadPool,
+                functions,
+                ramAccountingContext);
+
+        ListenableFuture<Bucket> result;
+        try {
+            result = fetchOperation.fetch();
```

Comment:
imho a bit strange that .fetch can fail itself and then the future could also fail too